diktat-rules/src/test/kotlin/org/cqfn/diktat/ruleset/chapter3/TrailingCommaWarnTest.kt

```diff
@@ -0,0 +1,268 @@
+            LintError(17, 25, ruleId, "${TRAILING_COMMA.warnText()} after VALUE_PARAMETER: description: String", true),
+            rulesConfigList = getRulesConfig("valueParameter")
+        )
+    }
+
+    @Test
+    @Tag(WarningNames.TRAILING_COMMA)
+    fun `check parameters with optional type`() {
+        lintMethod(
+            """
+                fun foo() {
+                    val sum: (Int, Int, Int,) -> Int = fun(
+                        x,
+                        y,
+                        z // trailing comma
+                    ): Int {
+                        return x + y + x
+                    }
+                    println(sum(8, 8, 8))
+                }
+                """.trimMargin(),
+            LintError(5, 25, ruleId, "${TRAILING_COMMA.warnText()} after VALUE_PARAMETER: z // trailing comma", true),
+            rulesConfigList = getRulesConfig("valueParameter")
+        )
+    }
+
+    @Test
+    @Tag(WarningNames.TRAILING_COMMA)
+    fun `check indexing suffix`() {
+        lintMethod(
+            """
+                class Surface {
+                    operator fun get(x: Int, y: Int) = 2 * x + 4 * y - 10
+                }
+                fun getZValue(mySurface: Surface, xValue: Int, yValue: Int) =
+                    mySurface[
+                        xValue,
+                        yValue // trailing comma
+                    ]
+                """.trimMargin(),
+            LintError(7, 25, ruleId, "${TRAILING_COMMA.warnText()} after REFERENCE_EXPRESSION: yValue", true),
+            rulesConfigList = getRulesConfig("referenceExpression")
+        )
+    }
+
+    @Test
+    @Tag(WarningNames.TRAILING_COMMA)
+    fun `check lambda parameters`() {
+        lintMethod(
+            """
+                fun main() {
+                    val x = {
+                            x: Comparable<Number>,
+                            y: Iterable<Number> // trailing comma
+                        ->
+                        println("1",)
+                    }
+                
+                    println(x,)
+                }
+                """.trimMargin(),
+            LintError(4, 29, ruleId, "${TRAILING_COMMA.warnText()} after VALUE_PARAMETER: y: Iterable<Number>", true),
+            rulesConfigList = getRulesConfig("valueParameter")
+        )
+    }
+
+    @Test
+    @Tag(WarningNames.TRAILING_COMMA)
+    fun `check when entry`() {
+        lintMethod(
+            """
+                fun isReferenceApplicable(myReference: KClass<*>) = when (myReference) {
+                    Comparable::class,
+                    Iterable::class,
+                    String::class // trailing comma
+                        -> true
+                    else -> false
+                }
+                """.trimMargin(),
+            LintError(4, 21, ruleId, "${TRAILING_COMMA.warnText()} after WHEN_CONDITION_WITH_EXPRESSION: String::class", true),
```

Comment:
Do when conitions always have this type? OR can they be 'without expression'?

Reply:
Yes, it can. Added tests